scripts/scons_helpers/build-3rdparty.py

```diff
@@ -664,18 +664,164 @@ def splitenv(envlist):
         env[k] = v
     return env
 
+# Guess platform argument for OpenSSL's Configure script basing on toolchain string.
+# (see `./Configure LIST` for a list of all platforms and their format)
+# See also openssl/Configurations/10-main.conf
+def openssl_get_platform(toolchain):
+    """
+    >>> openssl_get_platform('x86_64-pc-linux-gnu')
+    'linux-x86_64'
+    >>> openssl_get_platform('aarch64-linux-gnu')
+    'linux-aarch64'
+    >>> openssl_get_platform('mips-openwrt-linux-uclibc')
+    'linux-mips32'
+    >>> openssl_get_platform('arm-linux-gnueabihf')
+    'linux-generic32'
+    >>> openssl_get_platform('arm-bcm2708hardfp-linux-gnueabi')
+    'linux-generic32'
+    >>> openssl_get_platform('mips64-linux-musl')
+    'linux64-mips64'
+    >>> openssl_get_platform('arm-linux-androideabi')
+    'android-arm'
+    >>> openssl_get_platform('aarch64-linux-android')
+    'android64-aarch64'
+    >>> openssl_get_platform('i686-linux-android')
+    'android-x86'
+    >>> openssl_get_platform('x86_64-linux-android')
+    'android64-x86_64'
+    >>> openssl_get_platform('i686-pc-linux-gnu')
+    'linux-x86'
+
+    # other edge cases
+    >>> openssl_get_platform('android64')
+    'android64'
+    >>> openssl_get_platform('')
+    'cc'
+    """
+
+    # we assume that:
+    # - 'android' has more priority than 'linux'
+    # - if toolchain contains '64' and 'android' => we have android64 on 64-bit CPU
+    # - if toolchain contains '64' and 'linux' => we have linux64 on 64-bit CPU
+
+    bitness = 64 if '64' in toolchain else 32
+
+    # normalize os: make toolchain to have at most 3 components
+    toolchain = toolchain.\
+        replace('bcm2708hardfp-linux', 'linux').\
+        replace('openwrt-linux', 'linux').\
+        replace('pc-linux', 'linux')
+    # normalize arch: no suffix means 32 bit
+    toolchain = toolchain.replace('mips-', 'mips32-')
```

Comment:
The presence of this block indicates that heuristics in this function are not generic enough.

Ideally, the function should work without *knowing* if these four specific cases; these strings should be present only in tests.

* `bcm2708` is the name of random chip; we can't know and list names of all chips, and vendors can include chip names in toolchain;
* `openwrt` is the name of random distro; again, we can't list names of all of them;
* `pc-linux` is rather generic, but there are some variations, e.g. on some systems it's `uknown-linux`
* instead of `mips-`, it can be `mipsle-`, for example

It seems that all these problems are caused by relying on presence of specific tokens in specific places, like `linux` or `mips` before hyphen.

It seems all of them can be easily worked around by switching to a more relaxed matching. We can just check presence of `linux` and `mips` in toolchain string, and check presence or absence of other keywords like `64` or `armeabi`.

The lesser we assume about toolchain structure, the better...

Reply:
fair enough; rewritten most of this